sbt-converter/src/main/scala/org/scalablytyped/converter/internal/PluginRemoteCache.scala

```diff
@@ -47,6 +52,64 @@ object PluginRemoteCache {
             }
           }
           logger.warn(s"pushed to $pushLocation")
+
+        case RemoteCache.S3(pushConfiguration, _) =>
+          val (runCachePath, output) = inTask.value
+          val logger = WrapSbtLogger.task.value
+
+          val s3 = {
+            implicit class RichBuilder(s3ClientBuilder: S3ClientBuilder) {
+              def add[T](value: Option[T], builder: S3ClientBuilder => T => S3ClientBuilder) =
+                value.fold(s3ClientBuilder)(builder(s3ClientBuilder))
+            }
+
+            S3Client.builder()
+              .add(pushConfiguration.region, _.region)
+              .add(pushConfiguration.endpointOverride, _.endpointOverride)
+              .add(pushConfiguration.credentialsProvider, _.credentialsProvider)
+              .build()
+          }
+
+          def upload(relPath: RelPath)(implicit wd: Path) = {
+            val key = s"${pushConfiguration.prefix.getOrElse("")}/${relPath.toString()}"
+
+            val exists = Try(s3.headObject(HeadObjectRequest
+              .builder()
+              .bucket(pushConfiguration.bucket)
+              .key(key)
+              .build()
+            )).isSuccess
+
+            if (exists) {
+              logger.warn(s"skipping already cached file $relPath")
+            } else {
+              logger.warn(s"uploading $relPath")
+              s3.putObject(
+                PutObjectRequest
+                  .builder()
+                  .bucket(pushConfiguration.bucket)
+                  .key(key)
+                  .build(),
+                relPath.resolveFrom(wd).toNIO
+              )
+            }
+          }
+
+          Lock.synchronized {
+            locally {
+              implicit val wd = Utils.IvyLocal.value
+
+              output.allRelPaths.foreach(upload)
+            }
+            // this is somewhat convoluted. we know `wc` is a prefix of `runCachePath`, but compute `relPath`
```

Comment:
I hope I got it right :)

Reply:
Improving performance might make sense here, a few thoughts:
- for the antd project it takes 3 seconds to check whether those ~100 artifacts exist from my home dsl
- unfortunately there's neither a "uploadOnlyIfNotExists", nor a "batchExists"
- we could list all objects before uploading, but this is slightly trickier to implement as it uses paging and it might lead to actually worse performance if there are only few artifacts to be pushed, but actually thousands exist in the cache

Considering that, I think improving by using parallelism might be the best choice for now. wdyt?

Reply:
I've implemented the exists check and uploads using the async api so it's done in parallel. The only downside to this solution is that I had to add another dependency on https://github.com/scala/scala-java8-compat for easier interoperability with the java CompletableFuture. But as it by itself does not have any compiletime dependencies it should be fine?